### DISCUSSION TITLE
Void functons in Mojo

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#414] [Q&A] [answered]
Is Mojo going to implement void functions, which does not return anything? (as opposite to Python which always return None)

**Comment by gryznar** [comment]
I am asking cause it may be reasonable to indicate that `fn do_sth() - > Nome: return None` returns None by intention, not by assumption that everything returns.

Or `void` may be just default behaviour in `fn` for missing return type and trying to assign result of call to such fuction could raise compilation error, the same applies to presence of return.

In my opinion it would be cool if `fn` / `strict` won't require any inferring job. No type / type incompatible with declared - compilation error. It this scenario, no return type would mean just void func :)

**Comment by ksandvik** [comment]
To keep Mojo Pythonic it would be nice to use None rather than void. I doubt there's any optimization issues.

**Reply by gryznar** [reply]
But it is possible to return None directly ;)
I just want to distinguish situation in which Mojo `fn` does not return anything and in which it returns `None` :)

So pythonic void `fn` could be just without any return type. Return type as None means that None is returned via `return`. This "void" `fn` should be only enforced by the compiller instead of inferring `None` as return type :)

**Reply by gryznar** [reply]
Of course `def` should return `None`, but `fn` could be just more stricter and if no return is specified it shall assume, that function is void.

**Reply by gryznar** [reply]
They may be some viable usecases for such distinction and that's why void are in other languages

**Comment by gryznar** [comment]
> To keep Mojo Pythonic it would be nice to use None rather than void. I doubt there's any optimization issues.

Ahh, you mean a keyword instead of return type?

**Reply by gryznar** [reply]
How it be spelled it is not as important for me. Just to be clear in returning None vs not returning at all

**Comment by lattner** [comment]
I'm not exactly sure what you're suggesting - we'd like to avoid introducing the word/concept "void" because Python and Mojo already have both () and None which are similar but different.  It would be nice to avoid having yet-another similar but different thing.

Alternatively are you asking how the implementation works?  Right now it is convenient in the parser's implementation that functions always return exactly one thing, and that one thing can be None for the empty case or a tuple for the n-ary case.  The underlying compiler infrastructure we built correctly supports arbitrary # result values in its ABI and we do some amount of flattening as part of code generation so we can return multiple values in registers.

**Reply by gryznar** [reply]
I am just asking if Mojo will support (via compiler check and error) pure void concept in Mojo's `fn` (not returning). 

This support will be very useful to distinguish between no return (void) and return of `None` type. If I am not wrong proper type for not returning at all and returning `None` are the same as in Python:

```python
def return_none() -> None:
    return None
```

```python
def void_1 -> None:  # implicit return of None
    pass
```

```python
def void_2 -> None: # implicit return of None
    return
```

` -> None` is overloaded in Python and I hope that Mojo's type system and compiler will be suitable to allow user to distinguish not returning and returning `None` | `()` explicit.

**Reply by gryznar** [reply]
For me as a annotation freak this ambiguity is not the best part of Python because it produces confusion in few cases. Mojo could just do it better  in `fn` :)

In Python there is a concept of `typing.NoReturn` | `typing.Never` but this is only for cases in which interpreter cannot properly go out from function (without error for example), so unfortunatelly only `-> None` remains for functions which does not return anything explicitly.

As I proposed before it may be just no return type. No return type could mean void (and it would be enforced by the compiler if there is explicit return of value).

Here "having yet-another similar but different thing" could be very profitable for system programmers (and not only) in terms  of readability and having control (Lex Fridman would be happy 😁).

**Comment by lattner** [comment]
We currently treat "return" as a synonym for "return None".

**Comment by ksandvik** [comment]
Having None and void also would cause confusion, what to use where and why, where the end result is the same. You could always ignore checking for None.

**Comment by gryznar** [comment]
Hmm, I am just wondering if the compiler would be designed with concept of void function in mind, will be that profitable in terms of speed?

I mean from the compiler side what is better? 
1. Function which always return something (None or error)? 
2. void function

If 2 option is the answer, there may be reasonable to consider this concept?
If I am not wrong for void function there is no need to obtain and move returned data, so it may be faster.

**Reply by lattner** [reply, marked as answer]
The compiler already lowers return of None to the equivalent of void. :). There is no reason to complicate the user model here.  Thanks!

**Reply by gryznar** [reply]
Thanks for taking the time to clarify!
